src/backend/distributed/commands/alter_table.c

```diff
@@ -1572,3 +1577,76 @@ ExecuteQueryViaSPI(char *query, int SPIOK)
 		ereport(ERROR, (errmsg("could not finish SPI connection")));
 	}
 }
+
+
+/*
+ * SwitchToSequentialAndLocalExecutionIfRelationNameTooLong generates the longest shard name
+ * on the shards of a distributed table, and if exceeds the limit switches to sequential and
+ * local execution to prevent self-deadlocks.
+ *
+ * In case of a RENAME, the relation name parameter should store the new table name, so
+ * that the function can generate shard names of the renamed relations
+ */
+void
+SwitchToSequentialAndLocalExecutionIfRelationNameTooLong(Oid relationId,
+														 char *relationName)
+{
+	char *longestPartitionName = NULL;
+	if (!IsCitusTable(relationId))
+	{
+		return;
+	}
+
+	if (PartitionedTable(relationId))
+	{
+		longestPartitionName = LongestPartitionName(relationId);
+		if (longestPartitionName == NULL)
+		{
+			/* no partitions have been created yet */
+			return;
```

Comment:
Does a partitioned table without partitions not have an array type?

Reply:
They actually have types. Fixed it in a separate commit.

I copied this logic from `SwitchToSequentialAndLocalExecutionIfIndexNameTooLong`. I confirmed that we do not need to update that function with a similar logic as in here.

The index names of partitioned table shards are managed by Citus, so that's ok. However partition index names are created by vanilla PG, that actually causes deadlocks. So it's ok to skip partitioned tables with no partitions in `SwitchToSequentialAndLocalExecutionIfIndexNameTooLong`.

@onderkalaci Can you confirm if my deductions are correct here?

Reply:
> So it's ok to skip partitioned tables with no partitions in SwitchToSequentialAndLocalExecutionIfIndexNameTooLong.

This function seems to ignores the check if there are no partitions:

```C
	if (longestPartitionName == NULL)
	{
		/* no partitions have been created yet */
		return NULL;
	}
```

So, I guess you are right, but worth adding a regression test

Reply:
I no longer skip partitioned tables with no partitions.

All distributed tables check their longest shard name lengths, and in case it is a partitioned table then we check the names of the partition shards.

Overall, I think the tests cover this scenario, and I no longer need to add a test for a partitioned table with no partitions. Do you agree @onderkalaci? 

Note that this logic is partially broken for index statements, but I did not fix it here. Maybe I should create an issue for that.

Reply:
> Note that this logic is partially broken for index statements, but I did not fix it here. Maybe I should create an issue for that.


yes, please open an issue and we may want to fix it before Citus 10. Should be a quick fix on top of this?

Reply:
https://github.com/citusdata/citus/issues/4736